src/diagnostics.md

```diff
@@ -321,12 +322,27 @@ warnings, errors, fatal errors, suggestions, etc.
 [parsesses]: https://doc.rust-lang.org/nightly/nightly-rustc/rustc_session/parse/struct.ParseSess.html
 [session]: https://doc.rust-lang.org/nightly/nightly-rustc/rustc_session/struct.Session.html
 
+Diagnostics can be implemented as types which implement the `SessionDiagnostic`
+trait, this is preferred for new diagnostics as it enforces a separation
+between diagnostic emitting logic and the main code paths. For less-complex
+diagnostics, the `SessionDiagnostic` trait can be derived -- see [Diagnostic
+structs][diagnostic-structs]. Within the trait implementation, the APIs
+described below can be used as normal.
+
+[diagnostic-structs]: ./diagnostics/diagnostic-structs.md
+
```

Comment:
This paragraph feels like it was dropped in the wrong place... the paragraph that follows looks like a continuation of the previous one.

Reply:
Fixed, thanks! It was one paragraph lower than it ought to have been.